src/irc/mod.rs

```diff
@@ -46,94 +48,105 @@ pub struct IrcUserConnection<S: AsyncRead + AsyncWrite> {
+            Err(_) => return false,
+        };
+
+        match new_item {
+            IrcCommand::Nick { nick } => self.nick = Some(nick),
+            IrcCommand::User { user, real_name } => {
+                self.user = Some(user);
+                self.real_name = Some(real_name);
+            }
+            IrcCommand::Pass { password } => self.password = Some(password),
+            _c => {
+                //debug!(ctx.logger, "Ignore command during login"; "cmd" => c.command());
+            }
+        }
+        self.is_complete()
+    }
+}
+
+#[derive(Debug, Clone, Default)]
+struct UserNick {
+    nick: String,
+    user: String,
+    real_name: String,
+    password: String,
+}
+
+impl<S> IrcUserConnection<S>
+where
+    S: AsyncRead + AsyncWrite + 'static,
+{
     /// Given an IO connection, discard IRC messages until we see both a USER and NICK command.
-    pub fn await_login(
+    pub async fn await_login(
         server_name: String,
         stream: S,
         ctx: ConnectionContext,
-    ) -> Box<dyn Future<Item = IrcUserConnection<S>, Error = io::Error>> {
+    ) -> Result<IrcUserConnection<S>, io::Error> {
         trace!(ctx.logger, "Await login");
         let irc_conn =
             transport::IrcServerConnection::new(stream, server_name.clone(), ctx.clone());
 
         let ctx_clone = ctx.clone();
 
-        let f = StreamFold::new(irc_conn, UserNick::default(), move |cmd, mut user_nick| {
-            trace!(ctx.logger, "Got command"; "command" => cmd.command());
-            match cmd {
-                IrcCommand::Nick { nick } => user_nick.nick = Some(nick),
-                IrcCommand::User { user, real_name } => {
-                    user_nick.user = Some(user);
-                    user_nick.real_name = Some(real_name);
-                }
-                IrcCommand::Pass { password } => user_nick.password = Some(password),
-                c => {
-                    debug!(ctx.logger, "Ignore command during login"; "cmd" => c.command());
-                }
-            }
+        let folder = crate::stream_fold::StreamFold::new(irc_conn, UserNickBuilder::default());
+        (&folder).await;
 
-            (
-                user_nick.nick.is_some()
-                    && user_nick.user.is_some()
-                    && user_nick.real_name.is_some(),
-                user_nick,
-            )
-        })
-        .and_then(move |opt| {
-            if let Some((user_nick, mut irc_conn)) = opt {
-                info!(ctx_clone.logger, "Got nick and user");
-
-                let nick = user_nick.nick.expect("nick");
-
-                let password = if let Some(p) = user_nick.password {
-                    p
-                } else {
-                    irc_conn.write_password_required(&nick);
-                    return Err(io::Error::new(
-                        io::ErrorKind::InvalidData,
-                        "IRC user did not supply password",
-                    ));
```

Comment:
This did not get ported since I check that all fields are filled before exiting the `StreamFold`, whereas this version allows the password field to stay `None` until later